### ISSUE TITLE
Non-cascading application delete orphans ownership annotations

### ISSUE THREAD

**Issue by tgolsson** [argoproj/argo-cd#8179]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`

**Describe the bug/To Reproduce**

* I had a monolithic application consisting of multiple pods, services, etc, deployed all at once. 
* I changed my deployment flow to have five different deploy targets with identical cluster state
* I created an app-of-apps with all of these deploy targets as individual applications
* After creating the new top-level app I got warning about shared resources.
* I then deleted the old monolithic app, non-cascading (didn't want to disrupt users)
* Despite the old application being gone I still got SharedResourcesWarning because the application annotations were still left on all resources.
* To resolve this I had to manually deploy all my targets to clear the annotations, after which Argo could take over again.

**Expected behavior**

* A non-cascading delete removes ownership annotations from resources and/or a warning and/or a choice.(
OR
* A manual sync can opt-in to forcibly replace ownership -- I'd expect this to work with force or replace, but neither resolved it.

**Screenshots**

**Version**

```json
{
    "Version": "v2.2.2+03b17e0",
    "BuildDate": "2022-01-01T06:27:52Z",
    "GitCommit": "03b17e0233e64787ffb5fcf65c740cc2a20822ba",
    "GitTreeState": "clean",
    "GoVersion": "go1.16.11",
    "Compiler": "gc",
    "Platform": "linux/amd64",
    "KsonnetVersion": "v0.13.1",
    "KustomizeVersion": "v4.2.0 2021-06-30T22:49:26Z",
    "HelmVersion": "v3.7.1+g1d11fcb",
    "KubectlVersion": "v0.22.2",
    "JsonnetVersion": "v0.17.0"
}
```

**Comment by jannfis** [comment]
A sync of another application that manages the same resources should replace ownership. Before you sync, you will receive a shared ownership warning because Argo CD think it is still managed by the old (removed) Application.

Is the behavior you're seeing different from that?

**Comment by tgolsson** [comment]
Yes, when we debugged this we tried multipl combinations of sync settings and they all spun infinitely on syncing with the warnings showing. I also reproed the same later after renaming some applications, both for the k8s resources and within an app-of-app setup.

**Comment by jannfis** [comment]
By any chance, is one of the new apps (including the parent) having the same name as the old app you've deleted?

**Comment by tgolsson** [comment]
Nope, unfortunately not. Originally I had `a.b.c.app`. I then split that into `a.b.c.{d, e, f, g, h}` with `a.b.c.app-live` as parent, with the same resources. Then I renamed all (to get rid of the long names) to `xx.{d, e, f, g, h}` with parent `xx.bundle.app`.

The resources of the original `app` was split among `{d, e, f, g, h}`, and when renaming they retained mostly the same resources (some were added).

For full disclosure, in the second step (renaming) I took less precautions with turning off auto-sync etc (assuming the same issues would occur so why waste time on diligence..), so there might've been multiple syncs/applies here from the "master" app creating the new parent, the old parent creating the new children, and the new parent creating the new children.

**Comment by jannfis** [comment]
It sounds that you indeed might have shared resources between apps?

I just tested:

* Create an app `guestbook-one` in namespace `guestbook-test` with resources from the example guestbook app
* Sync app
* Delete app `guestbook-one` in a non-cascading manner
* Create new app `guestbook-two` with same settings as previously `guestbook-one`

App `guestbook-two` is out of sync from the start and I receive shared resource warning in application `guestbook-two`, telling me that the resources are also owned by `guestbook-one`. Once I sync `guestbook-two` (without any special sync options), application `guestbook-two` becomes in sync and the shared resource warning disappear.

**Comment by tgolsson** [comment]
That sounds fairly similar to what I did with different outcomes. I can for sure say that the shared resource warning said it was also owned by an app *that didn't exist*, i.e., *a.b.c.app*. The URL for that said it didn't exist, and it didn't show as an application in the frontend. I'm *not sure* whether it existed as a non-synced/dead resource inside the master app-of-apps.

**Comment by tgolsson** [comment]
Oh, sorry @jannfis , I know why we're getting different results! We specify `- FailOnSharedResource=true` in our common application config. So that's why we're getting this because it *can't* sync. I still think it's a *bug* because it's not actually shared anymore, we've deleted the other app - i.e, the ownership annotations are outdated.

**Comment by jannfis** [comment]
@tgolsson Argo CD at that point wouldn't know that resource is _not_ shared anymore. For example, it could be managed by another Argo CD instance on the cluster, e.g. due to a misconfiguration. And I think it's fair to assume that with that option set to `true`, the sync fails once a resource has the application instance label set on it.

I could see a possibility to be less strict for `FailOnSharedResource`. We could provide a third config value (next to `true` or `false`) that laxes the failure only for shared resources where the application claiming to own a resource actually exists in the current Argo CD installation.

**Comment by tgolsson** [comment]
I think removing the annotation is always safe, no matter whether there's multiple ArgoCD instances -- the annotation is the lock, no matter who owns the application manifest itself - right?

Let's assume *all* applications have the flag; then:

* Since the flag was set, we know that when deleting an application, it *must* be the true owner of any resource which currently has the annotation matching the application name - otherwise it would have been blocked from managing it in the first place! 
* Similarly, if it's *not* a match, then it *cannot* remove the annotation because it's *not* the manager of that resource, and thus it's not orphaning it in the first place.

If one of them has it set; presumably the application *without* the flag set violently takes ownership of the resources when deploying. Then, the "strict" application cannot be synced/resynced, while the non-strict application can.

* Deleting the non-strict application is *still* fine and safe because it *is* the true manager and it *should* remove the annotations. 
* Deleting the strict application is *still* fine because it *is not* the true manager and *should not* delete the annotations.

And by extension, because strict vs non-strict is really a feature of *acquiring* ownership, not *removing ownership*, I think it holds for if neither app is strict as well.

**Comment by jppitout** [comment]
We've run into the same problem. 

Unfortunately `FailOnSharedResource=true` doesn't play nice when renaming applications. I expected  `FailOnSharedResource=true` be smart enough to know when one of the applications was deleted and/or renamed and then set the label `argocd.argoproj.io/instance=<New Application Name>` accordingly. 

Instead we had to script an update of this label across all resources provisioned by the application :(

**Comment by Pastorpaul5432** [comment]
I am Pastor Paul of Christ Church Orphanage. We urgently need new adopters to provide a loving family. People from all walks of life can adopt children, but in particular we are looking for adoptive parents who will be able to meet a child’s needs. And to those of you looking for a baby to adopt, there are children; twins, boys and girls here. Like all families, adoptive families come in all shapes and sizes. Interested couples looking forward to adopting the babies should please contact us now. So if you’re thinking about starting the adoption process then please contact us now at: [christchurchorphanage@gmail.com](mailto:christchurchorphanage@gmail.com)  and you can also give your donations and charity to the orphanage. There is no need for these children to suffer any more so PLEASE HELP by coming to take them for adoption or donating to the orphanage. Help the little children. Thank you very much. God bless you as you come.
